### ISSUE TITLE
Build failures when building protobuf, grpc and opentelemetry from source

### ISSUE THREAD

**Issue by VivekSubr** [open-telemetry/opentelemetry-cpp#2192]
I build and install protobuf and grpc from source, on ubuntu 
https://github.com/protocolbuffers/protobuf
https://github.com/grpc/grpc

And then, I try to build opentelemetry,
```
cmake .. -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_STANDARD=20 -DBUILD_SHARED_LIBS=ON -DWITH_ABSEIL=ON -DWITH_STL=ON -DWITH_OTLP_GRPC=ON -DBUILD_TESTING=OFF 
    
cmake --build . --target all
```

And keep hit this linking error,
```
/usr/bin/c++ -rdynamic CMakeFiles/example_otlp_grpc.dir/grpc_main.cc.o -o example_otlp_grpc  -Wl,-rpath,/home/vivek/Client-Server/module/opentelemetry-cpp/build/examples/common/foo_library:/home/vivek/Client-Server/module/opentelemetry-cpp/build/exporters/otlp:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/trace:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/metrics:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/resource:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/common:/home/vivek/Client-Server/module/opentelemetry-cpp/build ../common/foo_library/libcommon_foo_library.so ../../exporters/otlp/libopentelemetry_exporter_otlp_grpc.so ../../exporters/otlp/libopentelemetry_otlp_recordable.so ../../sdk/src/trace/libopentelemetry_trace.so ../../sdk/src/metrics/libopentelemetry_metrics.so ../../sdk/src/resource/libopentelemetry_resources.so ../../exporters/otlp/libopentelemetry_exporter_otlp_grpc_client.so ../../sdk/src/common/libopentelemetry_common.so /usr/lib/x86_64-linux-gnu/libabsl_bad_variant_access.a /usr/lib/x86_64-linux-gnu/libabsl_bad_any_cast_impl.a /usr/lib/x86_64-linux-gnu/libabsl_city.a /usr/lib/x86_64-linux-gnu/libabsl_base.a /usr/lib/x86_64-linux-gnu/libabsl_raw_logging_internal.a /usr/lib/x86_64-linux-gnu/libabsl_log_severity.a /usr/lib/x86_64-linux-gnu/libabsl_spinlock_wait.a -lrt 

../../libopentelemetry_proto_grpc.so ../../libopentelemetry_proto.so /usr/lib/x86_64-linux-gnu/libprotobuf.so
/usr/bin/ld: ../../libopentelemetry_proto.so: undefined reference to `google::protobuf::Message::MaybeComputeUnknownFieldsSize(unsigned long, google::protobuf::internal::CachedSize*) const'
```

Searching for this symbol, 'MaybeComputeUnknownFieldsSize', in libprotobuf.so shows no results.
`vivek@LAPTOP-7E1991MU:~/Client-Server/module/opentelemetry-cpp/build$ nm -gD --demangle=auto /usr/lib/x86_64-linux-gnu/libprotobuf.so | grep -i "MaybeComputeUnknownFieldsSize"`

So... looks like opentelemetry latest needs some specific version of protobuf? The docs don't mention much: https://github.com/open-telemetry/opentelemetry-cpp/blob/main/docs/dependencies.md

Could someone comment on what might be going wrong?

**Comment by VivekSubr** [comment]
@owent, well I have a completely new ubuntu 22.04 VM where I do,

```
sudo apt install libprotobuf-dev
sudo apt install libabsl-dev
```

And try to build this repo and I get error,

```
ivek@LAPTOP-7E1991MU:~/Client-Server/module/opentelemetry-cpp/build$ cmake .. -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_STANDARD=20 -DBUILD_SHARED_LIBS=ON -DWITH_ABSEIL=ON -DWITH_STL=ON -DWITH_OTLP_GRPC=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=$ROOT/module/opentelemetry-cpp
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for architecture ARCH=x64
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
Building with standard library types...
PATH:/home/vivek/Client-Server/module/opentelemetry-cpp/examples/grpc/protos:/home/vivek/Client-Server/module/opentelemetry-cpp/examples/grpc/protos/messages.proto
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits - Success
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations - Success
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter - Success
Proto
CMake Error at examples/grpc/CMakeLists.txt:39 (target_include_directories):
  target_include_directories called with incorrect number of arguments
```

I think build should check whether it has all the protobufs it needs and throw some relevant error message at least?

**Comment by owent** [comment]
gRPC is also required to use OTLP gRPC exporter.
Could you try to use `sudo -E ./ci/setup_ci_environment.sh`, `sudo -E ./ci/install_protobuf.sh` and `sudo ./ci/setup_grpc.sh` to instal dependencies? They are also used in CI jobs.

**Comment by VivekSubr** [comment]
setup_grpc.sh fails for me on a new VM, 

```
[ 53%] Building CXX object absl/synchronization/CMakeFiles/graphcycles_internal.dir/internal/graphcycles.cc.o
/tmp/grpc/third_party/abseil-cpp/absl/synchronization/internal/graphcycles.cc: In member function ‘void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)’:
/tmp/grpc/third_party/abseil-cpp/absl/synchronization/internal/graphcycles.cc:451:26: error: ‘numeric_limits’ is not a member of ‘st ’
  451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
```

Strange error as the g++ version is up to date,
```
vivek@LAPTOP-7E1991MU:~/Client-Server/module/opentelemetry-cpp$ g++-12 --version
g++-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0
```

**Comment by owent** [comment]
Sorry, the setup_grpc.sh has some problems with gcc 10 or upper, this problem is fixed in #2163 but not merged yet.
Could you please install grpc between v1.49 and v1.54 with the installed protobuf?
You can also install it after modify setup_grpc.sh and remove the codes in `if [[ "${gcc_version}" < "${gcc_version_for_new_grpc}" ]];` and then run it again.

**Comment by VivekSubr** [comment]
Thank you, @owent, using these ci scripts is great.
